### PR TITLE
Make `Resolved*` in `PackageGraph` value types

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Depends on https://github.com/apple/swift-package-manager/pull/7161. Unblocks macros cross-compilation fix in https://github.com/apple/swift-package-manager/pull/7118.

### Motivation:

Value types are easier to reason about as they prevent "spooky action at a distance" bugs, and can be easily made `Equatable`, `Hashable`, and most important of all `Sendable`.

### Modifications:

Converted `ResolvedTarget`, `ResolvedProduct`, and `ResolvedPackage` from `final class`es to `struct`s.

Renamed `underlyingTarget`, `underlyingProduct`, and `underlyingPackage` to just `underlying` to avoid tautological patterns like `target.underlyingTarget`.

### Result:

It's easier to apply changes in `PackageGraph`. Also, in my benchmarks done locally this makes package graph resolution consistently ~5% faster when resolving the package graph of SwiftPM's own `Package.swift`.

#### `main` branch:

```
╒══════════════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                           │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞══════════════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Time (system CPU) (ms)           │       7 │       8 │      11 │      14 │      15 │      18 │      18 │      25 │
├──────────────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ms)            │      12 │      13 │      16 │      19 │      20 │      23 │      23 │      25 │
├──────────────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (user CPU) (μs)             │    4688 │    4857 │    4890 │    4956 │    5107 │    5347 │    5347 │      25 │
├──────────────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (wall clock) (ms)           │     390 │     397 │     402 │     409 │     414 │     420 │     420 │      25 │
╘══════════════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛
```

#### This PR's branch:

```
╒══════════════════════════════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╤═════════╕
│ Metric                           │      p0 │     p25 │     p50 │     p75 │     p90 │     p99 │    p100 │ Samples │
╞══════════════════════════════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╪═════════╡
│ Time (system CPU) (ms)           │       6 │       8 │      10 │      13 │      16 │      17 │      17 │      26 │
├──────────────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (total CPU) (ms)            │      11 │      13 │      15 │      18 │      20 │      21 │      21 │      26 │
├──────────────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (user CPU) (μs)             │    4476 │    4567 │    4648 │    4796 │    4931 │    4998 │    4998 │      26 │
├──────────────────────────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Time (wall clock) (ms)           │     375 │     379 │     384 │     391 │     394 │     398 │     398 │      26 │
╘══════════════════════════════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╧═════════╛
```

### Future directions:

I'm unable to fully remove `Resolved*Builder` classes yet, as they rely on `Target` being a reference type and module aliasing relying on side effects of modifying instances of `Target`. In turn, `Target` is a class hierarchy, which we should convert to multiple value types that use composition in the future. In the meantime this also prevents us from making it, and all types that contain it, `Sendable`.

### CODE REVIEW

**Review by tomerd — APPROVED** [review]

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift [hunk 1]
```diff
@@ -136,7 +136,7 @@ public final class SwiftTargetBuildDescription {
 
     /// The swift version for this target.
     var swiftVersion: SwiftLanguageVersion {
-        (self.target.underlyingTarget as! SwiftTarget).swiftVersion
+        (self.target.underlying as! SwiftTarget).swiftVersion
```
[the remark this answers is not in the document]
Comment (MaxDesiatov) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: Turns out we're already doing this cast safely in `init`. I've added a stored property to preserve the result of the cast.

File: Sources/Build/BuildDescription/ClangTargetBuildDescription.swift [hunk 2]
```diff
@@ -27,7 +27,7 @@ public final class ClangTargetBuildDescription {
 
     /// The underlying clang target.
     public var clangTarget: ClangTarget {
-        target.underlyingTarget as! ClangTarget
+        target.underlying as! ClangTarget
```
[the remark this answers is not in the document]
Comment (MaxDesiatov) [reply Sources/Build/BuildDescription/ClangTargetBuildDescription.swift]: Addressed same way as `SwiftTarget`.

File: Sources/PackageModel/PlatformRegistry.swift [hunk 3]
```diff
@@ -31,6 +30,19 @@ public final class PlatformRegistry {
...
+            .iOS,
+            .tvOS,
+            .watchOS,
+            .visionOS,
+            .linux,
+            .windows,
+            .android,
+            .wasi,
+            .driverKit,
+            .openbsd
```
Comment (tomerd) [inline Sources/PackageModel/PlatformRegistry.swift:45]: nit: sort alphabetically

File: Sources/PackageLoading/PackageBuilder.swift [hunk 4]
```diff
@@ -1146,11 +1146,11 @@ public final class PackageBuilder {
...
 
     func buildConditions(from condition: PackageConditionDescription?) -> [PackageCondition] {
-        var conditions = [PackageCondition]()
+        var conditions: [PackageCondition] = []
 
         if let config = condition?.config.flatMap({ BuildConfiguration(rawValue: $0) }) {
-            let condition = ConfigurationCondition(configuration: config)
-            conditions.append(.configuration(condition))
+            let condition = PackageCondition(configuration: config)
+            conditions.append(condition)
```
Comment (tomerd) [inline Sources/PackageLoading/PackageBuilder.swift:1153]: nit: `conditions.append(.init(configuration: config))`

**Review by neonichu — APPROVED** [review]

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

**Pull request merged**